### PR TITLE
Completed Screwtape waves 1-4

### PR DESCRIPTION
**Pull request by AugleBoBaugles**

### CODE REVIEW

File: src/Node.java [hunk 1]
```diff
@@ -38,7 +39,24 @@ public Node(int value) {
...
+    for (int i = 0; i < list.size(); i++){
+      if(i == 0){
+        this.value = list.get(i);
+      } else {
+        Node newNode = new Node(list.get(i));
+        recentNode.next = newNode;
+        newNode.prev = recentNode;
+        recentNode = newNode;
+        
+      }
```
Comment (auberonedu) [inline src/Node.java:58]: Nice!

File: src/NodeTest.java [hunk 2]
```diff
@@ -42,8 +42,28 @@ void testListConstructorWithEmptyList() {
...
-  // TODO: Add at least one more test for list constructor that would be useful and cover new ground.
+  // Add test for list constructor when passed null list
+  @Test
+  void testListConstructorWithNullList() {
+    List<Integer> nullList = null;
+
+    assertThrows(
+      NullPointerException.class, 
+      () -> new Node(nullList), "Expected constructor to throw NullPointerException for a null list.");
+  }
```
Comment (auberonedu) [inline src/NodeTest.java:53]: Nice test! We could also directly do `new Node(null)` instead of creating a null list reference first.

File: src/NodeTest.java [hunk 3]
```diff
@@ -67,6 +87,38 @@ void testToListWithThreeValues() {
...
+    middle.prev = head;
+    middle.next = tail;
+    tail.prev = middle;
+
+    // Act
+    List<Integer> values = middle.toList();
+
+    // Assert
+    assertEquals(List.of(7, 3), values);
+  }
```
Comment (auberonedu) [inline src/NodeTest.java:123]: Smart!

File: src/ScrewtapeInterpreter.java [hunk 4]
```diff
@@ -105,9 +107,26 @@ public void moveTapePointerToTail() {
...
+        openBrackets.push(i);
+      } else if (program.charAt(i) == ']'){
+        if (openBrackets.isEmpty()){
+          throw new IllegalArgumentException("Too many open brackets.");
+        }
+        bracketPointers.put(i, openBrackets.peek());
+        openBrackets.pop();
+      }
+    }
+
```
Comment (auberonedu) [inline src/ScrewtapeInterpreter.java:124]: Great!

File: src/ScrewtapeInterpreter.java [hunk 5]
```diff
@@ -129,8 +148,57 @@ public Map<Integer, Integer> bracketMap(String program) {
...
+      // <
+      else if (currentChar == '<'){
+        if (tapePointer.prev == null){
+          Node newNode = new Node(0);
+          tapePointer.prev = newNode;
+          newNode.next = tapePointer;
+          tapeHead = newNode;
+        }
+        tapePointer = tapePointer.prev;
+      }
```
Comment (auberonedu) [inline src/ScrewtapeInterpreter.java:185]: Nice!

File: src/ScrewtapeInterpreter.java [hunk 6]
```diff
@@ -129,8 +148,57 @@ public Map<Integer, Integer> bracketMap(String program) {
    * @throws IllegalArgumentException If the program contains unmatched brackets.
    */
   public String execute(String program) {
-    // TODO: Implement this
     // If you get stuck, you can look at hint.md for a hint
-    return null;
+    String output = "";
```
Comment (auberonedu) [inline src/ScrewtapeInterpreter.java:152]: Consider using StringBuilder when repeatedly adding to a String.

File: src/ScrewtapeInterpreterTest.java [hunk 7]
```diff
@@ -26,6 +26,70 @@ void testNestedBracketMap() {
...
+    // Arrange
+    ScrewtapeInterpreter interpreter = new ScrewtapeInterpreter();
+    String program = "+-<<>>_+_+";
+
+    Map<Integer, Integer> expectedMap = new HashMap<>();
+
+    Map<Integer, Integer> actualMap = interpreter.bracketMap(program);
+
+    assertEquals(expectedMap, actualMap);
+  }
```
Comment (auberonedu) [inline src/ScrewtapeInterpreterTest.java:90]: Smart test!

**Review by auberonedu — COMMENTED** [review]
Awesome job!
